File: user-service/src/main/java/rs/edu/raf/security/JwtFilter.java
package rs.edu.raf.security;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
public class JwtFilter extends OncePerRequestFilter {
    private final CustomUserDetailsService userService;
    private final JwtUtil jwtUtil;

    public JwtFilter(CustomUserDetailsService userService, JwtUtil jwtUtil) {
        this.userService = userService;
        this.jwtUtil = jwtUtil;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
                                    FilterChain filterChain) throws ServletException, IOException {
        if (httpServletRequest.getServletPath().equals("/api/user/login")){
            filterChain.doFilter(httpServletRequest, httpServletResponse);
        } else {
            String jwt = jwtUtil.extractToken();
            String username = null;

            if (jwt != null) {
                if (jwt.startsWith("Bearer ")) jwt = jwt.substring(7);
                username = jwtUtil.extractUsername(jwt);

                if (username != null && SecurityContextHolder.getContext().getAuthentication() == null) {
                    UserDetails userDetails = this.userService.loadUserByUsername(username);

                    if (jwtUtil.validateToken(jwt, userDetails)) {

                        UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(
                                userDetails, null, userDetails.getAuthorities());
                        usernamePasswordAuthenticationToken
                                .setDetails(new WebAuthenticationDetailsSource().buildDetails(httpServletRequest));

                        SecurityContextHolder.getContext().setAuthentication(usernamePasswordAuthenticationToken);
                    }
                }
                filterChain.doFilter(httpServletRequest, httpServletResponse);
            }
            else {
                filterChain.doFilter(httpServletRequest, httpServletResponse);
            }
        }
    }
}

File: user-service-kotlin/src/main/kotlin/org/example/userservicekotlin/security/JwtFilter.kt
package org.example.userservicekotlin.security

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtFilter(val customUserDetailsService: CustomUserDetailsService, val jwtUtil: JwtUtil): OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        if (request.getServletPath() == "/api/user/login") {
            filterChain.doFilter(request, response)
        } else {
            var jwt = jwtUtil.extractToken()
            val username: String?

            if (jwt != null) {
                if (jwt.startsWith("Bearer ")) jwt = jwt.substring(7)
                username = jwtUtil.extractUsername(jwt)

                if (SecurityContextHolder.getContext().authentication == null) {
                    val userDetails: UserDetails = this.customUserDetailsService.loadUserByUsername(username)

                    if (jwtUtil.validateToken(jwt, userDetails)) {
                        val usernamePasswordAuthenticationToken = UsernamePasswordAuthenticationToken(
                            userDetails, null, userDetails.authorities
                        )
                        usernamePasswordAuthenticationToken.details =
                            WebAuthenticationDetailsSource().buildDetails(request)

                        SecurityContextHolder.getContext().authentication = usernamePasswordAuthenticationToken
                    }
                }
                filterChain.doFilter(request, response)
            } else {
                filterChain.doFilter(request, response)
            }
        }
    }
}